src/main/java/de/zalando/aruha/nakadi/service/subscription/zk/CuratorZkSubscriptionClient.java

```diff
@@ -0,0 +1,263 @@
+package de.zalando.aruha.nakadi.service.subscription.zk;
+
+import de.zalando.aruha.nakadi.service.subscription.SubscriptionWrappedException;
+import de.zalando.aruha.nakadi.service.subscription.model.Partition;
+import de.zalando.aruha.nakadi.service.subscription.model.Session;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.locks.InterProcessSemaphoreMutex;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CuratorZkSubscriptionClient implements ZkSubscriptionClient {
+    private static final String STATE_INITIALIZED = "INITIALIZED";
+    private static final Charset CHARSET = Charset.forName("UTF-8");
+
+    private final CuratorFramework curatorFramework;
+    private final InterProcessSemaphoreMutex lock;
+    private final String subscriptionId;
+    private final Logger log;
+
+    public CuratorZkSubscriptionClient(final String subscriptionId, final CuratorFramework curatorFramework) {
+        this.subscriptionId = subscriptionId;
+        this.curatorFramework = curatorFramework;
+        this.lock = new InterProcessSemaphoreMutex(curatorFramework, "/nakadi/locks/subscription_" + subscriptionId);
+        this.log = LoggerFactory.getLogger("zk." + subscriptionId);
+    }
+
+    @Override
+    public void runLocked(final Runnable function) {
+        log.info("Taking runLocked for " + function.hashCode());
+        try {
+            lock.acquire();
+            log.debug("Lock taken " + function.hashCode());
+            try {
+                function.run();
+            } finally {
+                log.info("Releasing runLocked for " + function.hashCode());
+                try {
+                    lock.release();
+                } catch (final Exception e) {
+                    log.error("Failed to release runLocked", e);
+                    throw e;
```

Comment:
If you get to `finally` block with an exception - you will loose it as you throw a new one.